Return position `impl Trait` in traits

tmandry opened the pull request:
* Permit `impl Trait` in fn return position within traits and trait impls.
* Allow `async fn` in traits and trait impls to be used interchangeably with its equivalent `impl Trait` desugaring.
* Allow trait impls to `#[refine]` an `impl Trait` return type with added bounds or a concrete type.

## Background

This RFC is a collaboration between myself and @compiler-errors, and is based on an [earlier RFC](https://github.com/rust-lang/rfcs/pull/3193) by @nikomatsakis.

The primary changes from that RFC are:

* `async fn` is now allowed to be used interchangeably with its equivalent `impl Trait` desugaring.
* `#[refine]` is now included as a way to add more information about the returned type in an impl, either by adding bounds to the `impl Trait` type, or by using a concrete type. #3245 did not exist at the time of the previous RFC (#3193), and was added partly in response to discussion there.
* You are now allowed to use `dyn Trait` on a trait with a method returning `impl Trait`, as long as that method has a `where Self: Sized` bound.

[Rendered](https://github.com/tmandry/rfcs/blob/rpitit/text/0000-return-position-impl-trait-in-traits.md)

nikomatsakis commented:
@rfcbot fcp merge

This feature has been a long time coming and feels like a no brainer to me. I'm going to start the merge proceedings.

nikomatsakis commented:
@rfcbot fcp cancel

@rustbot labels -T-types

nikomatsakis commented:
@rfcbot fcp merge

This feature has been a long time coming and feels like a no brainer to me. I'm going to start the merge proceedings.

spastorino commented on text/0000-return-position-impl-trait-in-traits.md:
````diff
@@ -0,0 +1,508 @@
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    #[refine]
````
Michael has a PR that implement these checks. So resolving this comment ...

lcnr commented:
Wrote this down after midnight so it's not as clear and concise as I would have liked. I only really thought about this in-depth after Niko proposed the FCP.

tl;dr: RPITIT is not a trivial extension of TAIT or RPIT and does not simply fall out of the desire for a consistent design. As [I stated separately](https://github.com/rust-lang/rust/pull/109010#issuecomment-1486836253), I don't think we should add return type notation **RTN**. This causes RPITIT to be a major footgun for library authors without a significant benefit.

I very much agree that we should add support to `async fn` in traits but am not yet sold that RPITIT is the right approach for that.

## RPITITs are not simple opaque types

[moved to a hackmd as it is somewhat pedantic](https://hackmd.io/@lcnr/H1WwQvxN3)

I don't agree that RPITIT will result in a more consistent language given that the semantics of RPITIT are significantly different from RPIT. It is easy to provide context to the error message when trying to use it, recommending an associated type with TAIT.

## RPITIT can result in worse library design

To keep this short, I prefer https://github.com/tmandry/rfcs/blob/rpitit/text/0000-return-position-impl-trait-in-traits.md#what-about-using-an-explicit-associated-type over RPITIT and don't think we should add RPITIT (at least for now).

The RFC uses `IntoIterator` as a good example for where RPITIT would be applicable. [Searching for `::IntoIter: ` on github](https://github.com/search?q=%3A%3AIntoIter%3A+&type=code) results in a non-negligible amount of crates which use bounds on `IntoIter`. Using RPITIT without a way to name the anonymous projection would have prevented this code. Given that I fairly strongly dislike RTN, this is a major issue with RPITIT imo.

Going through the list of cases where an explicit associated type should be used, we can reverse it to get requirements to correctly use RPITIT:

- it is irrelevant whether users provide concrete types as you can't expect them to always add `#[refine]`
- no user will ever need to constrain the `impl Trait` to a concrete value: [also not the case for `IntoIterator`](https://github.com/search?q=IntoIterator%3CIntoIter+%3D+&type=code)
- no user will ever need to add additional bounds for the `impl Trait`

We have proof that the second and third point do not hold for `IntoIterator` and I am certain that the first point also does not hold. I don't think RPITIT would be sensible for `IntoIterator`.

More generally, I cannot think of a trait where RPITIT would be appropriate given these requirements, unless the trait is local to your library or sealed. I assume that given the ability, library users will sometimes end up using RPITIT simply because it is possible, which will then require breaking changes to fix later on.

## RPITIT is not necessary for async functions in traits

By going with the "explicit associated type" route, we can instead add some attribute to express the concept of "this named associated type may be implicitly defined by this function", something like
```rust
trait IntoIterator {
    type Item;
    type IntoIter: Iterator<Item = Self::Item>;
    #[may_implicitly_define(IntoIter)]
    fn into_iter(self) -> Self::IntoIter;
}

impl<T: Iterator> IntoIterator for T {
    type Item = T::Item;
    // does not explicitly define `IntoIter`,
    fn into_iter(self) -> Self::IntoIter {
        self
    }
    // or
    fn into_iter(self) -> impl Iterator<Item = T::Item> {
        self
    }
} 
```
A similar approach can then be used for `async` functions as well. I don't yet know how exactly these attributes should look but I do prefer such attributes over RPITIT.

## RPITIT is at most a linear ergonomics improvement

As stated in the RFC, RPITIT should be equivalent to its desugaring using TAIT and associated types. This desugaring has some repetition but it should pretty much never significantly impact development speed or readability.

Given the fairly mechanical desugaring, we should be able to emulate RPITIT using TAIT and a set of macros for trait definitions and impls. If the manual desugaring poses a major issue, users should be drawn to such a set of macros, at which point we can reconsider adding RPITIT.

comex commented:
I agree with @lcnr that this feature effectively requires return type notation.  In some hypothetical future where Rust decided against return type notation, RPITIT would clearly not hold its weight.  Therefore it does seem strange to be landing it without at least talking about return type notation.

Unlike @lcnr, I think return type notation sounds great.  But I am confused about the details.  Is it going to be `()` or `(..)`?  (The [implementation PR](https://github.com/rust-lang/rust/pull/109010#issuecomment-1486836253) suggests it might switch over to `(..)` in the future, but [today's blog post](https://blog.rust-lang.org/inside-rust/2023/05/03/stabilizing-async-fn-in-trait.html) uses `()`.)  And since neither of those syntaxes actually lists the arguments, how does that interact with generic functions, where the return type might conditionally implement a trait depending on the argument types?

dhardy commented:
> We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.

Is deferring this acceptable? It seems like it should be a core part of the proposal. Then again, we still have no syntax for naming the return type of a closure or a regular `impl Trait` method, so perhaps not.

Should we at least have somewhere to discuss relevant issues:

- Syntax for naming the return type of RPITIT (or something more general like `return_type_of<F>`)
- Associated type aliases (which [are not associated type defaults](https://github.com/rust-lang/rfcs/pull/2532#issuecomment-465774140))

dhardy commented:
An alternative approach:
```rust
trait T {
    auto type Fut = return_type_of<Self::fun>;

    async fn fun(&self) -> i32;
}
```
`auto type` defines an associated type, except *it cannot be defined at the impl site*. (Optionally, we could allow impl-site definitions that correlate exactly with the inferred type in order to allow migration from assoc. types to assoc. auto types, and just warn that the definition is redundant.)

`async fn fun` (and any other usage of RPIT in a trait method) could be legal *only* given a corresponding `auto type`. By making a corresponding `auto type` *required*, we (a) always have a named type for usage in bounds, field types of generic structs, etc., and (b) are more transparent about why `dyn Trait` is not valid (at least not without constraining the assoc. auto type, if that is even possible).

Putting both of these together would allow e.g. migrating `ImplIterator::IntoIter` into an `auto type` and `into_iter` into an RPITIT method.

kpreid commented:
As a user of stable Rust, I'd like to see effort put first into stabilization of TAIT, which would enable many of the use cases that RPITIT would, and also provide more information about what benefits RPITIT would or would not provide and how it should be designed, since it results in a smaller delta (“write this trait using an associated type” vs. “can't practically write this trait at all”).

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+[summary]: #summary
+
+* Permit `impl Trait` in fn return position within traits and trait impls.
+* Allow `async fn` in traits and trait impls to be used interchangeably with its equivalent `impl Trait` desugaring.
```
One consequence of this is that it basically stabilizes the `impl Future` desugaring. I don't imagine we'll want to change it, and enough things depend on the `impl Future` desugaring now that we probably can't change it practically already, but it still means we're committing to what in some ways feel like internal implementation details to me.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+## Example use case
+
+The use case for `-> impl Trait` in trait functions is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:
```
This makes it sound like the primary benefit of RPITIT is ergonomics, rather than adding significant new expressiveness to the language. In other words, it makes some things easier to say but doesn't add the ability to say new things. Is that right?

dhardy replied:
Yes and no. For example, you can do this today:
```rust
trait Foo {
    type Fut<'a>: std::future::Future<Output = ()> + 'a
    where
        Self: 'a;
        
    fn fut(&self) -> Self::Fut<'_>;
}

// Foo::fut is essentially an async fn:
async fn with<F: Foo>(foo: &F) {
    foo.fut().await
}
```
But, as in the RFC, this is not dyn-safe:
```rust
// error[E0191]: the value of the associated type `Fut` (from trait `Foo`) must be specified
async fn with_dyn(foo: &dyn Foo) {
    foo.fut().await
}
```
However, *implementing* `Foo` as written above requires specifying the type of `Foo::Fut` (sometimes annoying, sometimes impossible without type-alias impl-trait), despite the fact that such a definition is entirely redundant with the definition of `Foo::fut`.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
````diff
@@ -0,0 +1,508 @@
+}
+```
+
+It can, however, give a more specific type with `#[refine]`:[^refine]
````
Is there a rationale given somewhere for requiring `#[refine]`?

nikomatsakis replied:
https://github.com/rust-lang/rfcs/pull/3245

tmandry replied:
I think the RFC includes the rationale and is linked to by the footnote, so resolving. Feel free to unresolve if you think there should be more explicit text here.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### async fn desugaring
+
+`async fn` always desugars to a regular function returning `-> impl Future`. When used in a trait, the `async fn` syntax can be used interchangeably with the equivalent desugaring in the trait and trait impl:
```
Presumably the reason for doing this is so that we can use `#[refine]` to make stricter promises on the impl? For example, adding a `Send` bound at the impl level.

tmandry replied:
That's one (another example of an added promise would be that you don't capture `self`). A whole other reason would be to include code that runs at call time instead of during `poll`.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
````diff
@@ -0,0 +1,508 @@
+}
+```
+
+```rust
+trait SomeTrait {
+    fn method<P0, ..., Pm>()
+}
+```
````
What is this example meant to show?

tmandry replied:
I think it was left by mistake, removing.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+    * The impl must use `impl Trait` syntax to name the corresponding type, and
+    * The return type in the trait must implement all bounds `I_0 + ... + I_n` specified in the impl return type. (Taken with the first outer bullet point, we can say that the bounds in the trait and the bounds in the impl imply each other.)
+
+[^refine]: `#[refine]` was added in [RFC 3245: Refined trait implementations](https://rust-lang.github.io/rfcs/3245-refined-impls.html). This feature is not yet stable.
```
Does this mean this RFC is dependent on 3245 being stabilized first, or is the plan to include a subset of 3245 in this RFC?

tmandry replied:
Added clarifying text to this footnote; it just means any code that requires `#[refine]` won't work until that feature is stabilized. Simple examples without refinement will continue to work.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
````diff
@@ -0,0 +1,508 @@
+    }
+}
+
+// Not OK:
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    fn into_iter(self) -> std::vec::IntoIter<u32> {
+        self.into_iter()
+    }
+}
+```
````
This example is Not OK because it's missing the `#[refine]` attribute, right?

tmandry replied:
Right. Added to the comment.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### Does auto trait leakage still occur for `-> impl Trait` in traits?
+
+Yes, so long as the compiler has enough type information to figure out which impl you are using. In other words, given a trait function `SomeTrait::foo`, if you invoke a function `<T as SomeTrait>::foo()` where the self type is some generic parameter `T`, then the compiler doesn't really know what impl is being used, so no auto trait leakage can occur. But if you were to invoke `<u32 as SomeTrait ...
```
Once we have associated return types / return type notation, I wonder if we can change this and if it'd be worth changing it? As long as it's not too burdensome to be explicit, that seems like it might be a better situation than the current state of things.

tmandry replied:
It _might_ be worth changing, but not as part of this RFC. If we ever remove auto trait leakage I think we should do it globally, not as part of individual new expansions of `-> impl Trait`, and this would of course require an edition migration.

We also aren't that close to removing it; we would need some kind of syntax that means "this type impls `Send` if `T: Send`" for some input type `T`, which we don't have, and that was a key part of the reason for auto trait leakage in the first place.

eholk commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### Can traits migrate from `impl Trait` to a named associated type?
+
+Generally yes, but all impls would have to be rewritten.
```
We'd probably need TAIT too, right?

nikomatsakis replied:
Good callout.

nikomatsakis replied:
I think it'd be worth expanding this answer to go into a bit more detail about *why*

nikomatsakis replied:
```suggestion
Generally yes, but all impls would have to be rewritten to include the definition of the associated type. In many cases, some form of type-alias impl trait (or impl trait in associated type values) would also be required. For example, if we changed the `IntoIntIterator` trait from the motivation to use an explicit associated type..

```rust
trait IntoIntIterator {
    type IntIter: Iterator<Item = u32>;
    fn into_iter(self) -> Self::IntIter;
}
```

...then impls like...

```rust
impl IntoIntIterator for MyType {
    fn into_int_iter(self) -> impl Iterator<Item = u32> {
        (0..self.len()).map(|x| x * 2)
    }
}
```

...would no longer compile, because they are not specifying the value of the `IntIter` associated type. Moreover, the value for this type would be impossible to express without `impl Trait` notation, as it embeds a closure type.

nikomatsakis commented:
Hey all, thanks for the comments. I want to summarize the points raised thus far:

* RPITITs are different from other "opaque" uses of `impl Trait` (e.g., RPIT, TAIT), in that they desugar, on the trait side, to an associated type (on the impl side, they are exactly analogous, I believe).  This weakens the "more consistent language" story, since it's not just a matter of applying the exact same desugaring we already have elsewhere.
* On their own, RPITITs work against maximum flexibility, as compared to normal associated types, because there is no way to add bounds to them. Using them in widely used libraries is likely to be a kind of anti-pattern. 
* If we stabilize TAITs or uses of `impl Trait` in associated type values (AssocIT), that -- combined with GATs -- lets users model RPITIT explicitly. It'd be a better place to start.
* There is a bit of latent thread in @eholk's comments about whether `#[refine]` is the right design, and I've had conversations about this with a few others (e.g., @compiler-errors). 
* lcnr points out that RPITIT is not *necessary* for async functions in traits, one can imagine other solutions.

nikomatsakis commented:
> On their own, RPITITs work against maximum flexibility, as compared to normal associated types, because there is no way to add bounds to them. Using them in widely used libraries is likely to be a kind of anti-pattern.

This is true, but it's also consistent with `impl Trait` elsewhere. For example, best practice for widely portable libraries remains to use newtypes rather than RPIT, so that people can name the types returned by your functions, for example for use in structs. Similarly, `impl Trait` in argument position is less flexible, as people cannot use turbofish. 

I don't love this about impl Trait; of course the plan has always been to stabilize additional features that will help people combine the convenience of impl Trait with the full explicitness. For example, TAITs allow you to name your return types. 

That brings me to this point, quoting @comex ...

> I agree with @lcnr that this feature effectively requires return type notation. In some hypothetical future where Rust decided against adding return type notation, RPITIT would clearly not hold its weight.

...I agree that we need a solution to some of the problems that @lcnr raised that will occur with RPITIT on its own, but I don't agree that this means RPITIT doesn't hold its weight if we don't add associated return types / return type notation. It's perhaps true that RPITIT doesn't hold its weight if we do *nothing* else, but there are other routes we could go to solve the shortcomings besides associated return types. 

The point of separating out *just* the base RPITIT into its own RFC is precisely to identify an MVP and enable us to make incremental progress without having to get full agreement on every aspect of the design. This is a time-honored Rust tradition, though it can be frustrating when those extensions get stuck or fail to make progress. 

For example, if we opt to go with the "TAIT-like" route, we might make a way to refactor a trait that uses `-> impl Trait` into one that uses a named associated type, much like we are talking about doing with TAIT, where you can factor an inherent method from using `-> impl Trait` to using a named type alias. This is pretty similar to what @lcnr later proposes with named associated types.

As it happens, I like the associated return types idea, but I don't think that adding impl Trait in traits means we *have* to accept associated return types. All we are committing to is that writing `-> impl Trait` in a trait winds up resolving to some kind of generic associated type, which I think is really the only thing it could possibly mean. 

The fact that, if you use it in a public library, your users may get stuck is true but not unique to this future, and also frankly pretty common with `impl Trait` to begin with. We should address it, but it will be easier to do that once the feature exists and we've had more experience using it, and I don't think it's going to change anything about the semantics of programs described by this specific RFC.

nikomatsakis commented:
On the final point...

> If we stabilize TAITs or uses of impl Trait in associated type values (AssocIT), that -- combined with GATs -- lets users model RPITIT explicitly. It'd be a better place to start.

...I agree it'd be great to stabilize those features, and I support that work. But to me it feels like a false comparison to say that accepting this RFC is taking away from making progress on those features. We're working on both simultaneously, and I feel like (as I wrote before), I believe we will want people to be able to write `-> impl Trait` in traits and to have it desugar to some kind of associated type, whatever route we take to get there.

Ultimately it seems like this is the core question, really: "I believe we will want people to be able to write `-> impl Trait` in traits and to have it desugar to some kind of associated type, whatever route we take to get there." If we do want that, I think we should take a minimal RFC that holds space for us to improve its shortcomings later. 

But maybe you don't believe that, and you think it'd be better to have people use explicit associated types and TAITs. I can see that argument. I don't agree with that tradeoff, but I can see why people would feel that a language that doesn't have `impl Trait` in traits is, on net, simpler.

dhardy commented:
> The point of separating out just the base RPITIT into its own RFC is precisely to identify an MVP and enable us to make incremental progress without having to get full agreement on every aspect of the design. This is a time-honored Rust tradition, though it can be frustrating when those extensions get stuck or fail to make progress.

Allowing progress on RPITIT without getting stuck bike-shedding an acceptable method of naming the elided types is perfectly acceptable in my opinion. But I still think we need a place to discuss that because it is an important to a lot of uses of the feature (and other `impl Trait` features).

There are a few approaches:

- Direct: add some syntax like `Trait::method()` or `return_type_of::<Trait::method>` or ... (the list is endless)
- Indirect with TAIT. This is... useful, but weird both in that there is no direct definition and in that one cannot add a type alias within all contexts (e.g. if a struct field or method parameter is type-dependant on an RPITIT method while being generic over the trait)
- Indirect with an inherent type alias in the trait (what I tried to do poorly above with `auto type`)
- Indirect via an associated TAIT on an extension trait?

nikomatsakis commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+Similarly, the equivalent `-> impl Future` signature in a trait can be satisfied by using `async fn` in an impl of that trait.
+
+## Nested impl traits
```
So, this example compiles today:

```rust
#![feature(return_position_impl_trait_in_trait)]

trait Foo {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug);
}

impl Foo for u32 {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug) {
        (22, 44)
    }
}
```

this makes sense, given the desugaring described in the RFC, but I didn't find any explicit mention of it (though upon re-reading I realize now it's implicit in sentences like "Each `-> impl Trait` notation appearing in a trait fn return type is effectively desugared to an anonymous associated type. "). We should add an explicit section on this, or perhaps generalize this one to "legal positions for `impl Trait` to appear".

tmandry replied:
I think the intent should be that anything that is legal in a free function should be legal here, so I'll update the section to say that.

tmandry replied:
@nikomatsakis Can you check that the updated text resolves your concerns?

nikomatsakis commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### Naming return types
+
+We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.
```
```suggestion
We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.
```

This sentences is ambiguous. At first I thought you meant "introducing a named associated type into the trait", but now I think you mean something like RTN? It'd be useful to reword it for clarity. In particular, I think what this means is "some way for generic code using a trait to place bounds on the return type of `-> impl Trait` functions"? (This is slightly different from being able to name the results of arbitrary `-> impl Trait` functions, e.g., if a function returns `-> (impl Trait, impl Trait)`.)

nikomatsakis commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### Can traits migrate from a named associated type to `impl Trait`?
+
+Not compatibly, no, because they would no longer have a named associated type.
```
```suggestion
Not compatibly, no, because they would no longer have a named associated type. The "future directions" section discusses the possibility of allowing users to explicitly give a name for the associated type created, which would enable this use case.
```

nikomatsakis commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+### Would there be any way to make it possible to migrate from `impl Trait` to a named associated type compatibly?
+
+Potentially! There have been proposals to allow the values of associated types that appear in function return types to be inferred from the function declaration. So the trait has `fn method(&self) -> Self::Iter` and the impl has `fn method(&self) -> impl Iterator`, then the impl would also be inferred to have `type Iter = impl Iterator` (and the return type rewritten to reference it). This may be ...
```
```suggestion
Potentially! There have been proposals to allow the values of associated types that appear in function return types to be inferred from the function declaration. So, using the example from the previous question, the impl for `IntoIntIterator` could infer the value of `IntIter` based on the return type of `into_int_iter`. This may be a good idea, but it is not proposed as part of this RFC.
```

tmandry replied:
This will be an incomplete solution once we support dyn dispatch because you'd have to avoid breaking uses of `dyn Trait` that don't specify the associated type. There's also a future possibilities section on this that needs updating.

nikomatsakis commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,508 @@
+
+The same is not true for `Iterator::Item`: because `Item` is so central to what an `Iterator` is, and because it rarely makes sense to use an opaque type for the item, it would remain an explicit associated type.
+
+# Prior art
```
the RFC should address the points raised by @lcnr 

```suggestion
## Should library traits migrate to use `impl Trait`?

Potentially, but not necessarily. Using `impl Trait` in traits imposes some limitations on generic code referencing those traits. While `impl Trait `desugars internally to an associated type, that associated type is anonymous and cannot be directly referenced by users, which prevents them from putting bounds on it or naming it for use in struct declarations. This is similar to `-> impl Trait` in free and inherent functions, which also returns an anonymous type that cannot be directly named. Just as in those cases, this likely means that widely used libraries should avoid the use of `-> impl Trait` and prefer to use an explicit named associated type, at least until some of the "future possibilities" are completed. However, this decision is best reached on a case-by-case basis: the real question is whether the bounds named in the trait will be sufficient, or whether users will wish to put additional bounds. In a trait like `IntoIterator`, for example, it is common to wish to bound the resulting iterator with additional traits, like `ExactLenIterator`. But given a trait that returns `-> impl Debug`, this concern may not apply.
```

digama0 commented on text/0000-return-position-impl-trait-in-traits.md:
````diff
@@ -0,0 +1,683 @@
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
````
Or `U32Iterator`

aliemjay reviewed (commented):
I'm concerned about an inconsistency with inherent impls regarding the capture rules for lifetimes. It can be shown in this example:
```rust
struct Foo<'a>(&'a str);

// ERROR: `impl Sized` doesn't capture 'a
impl<'a> Foo<'a> {
    fn into(self) -> impl Sized { self }
}

trait IntoImpl {
    fn into(self) -> impl Sized;
}

// OK! Allowed by the RFC.
impl<'a> IntoImpl for Foo<'a> {
    fn into(self) -> impl Sized { self }
}
```

>We say that a generic parameter is captured if it may appear in the hidden type. These rules are the same as those for -> impl Trait in inherent impls.

Given the difference above, this statement is no longer true.


Additionally, in the capture rules, the text talks only about generic parameters of the trait implementation - it says nothing about the trait definition. It is not clear to me, for example, whether `impl Sized` captures the lifetime `'a` here:
```rust
trait Trait<'a> {
    fn test() -> impl Sized;
}
```

tmandry commented:
@aliemjay Oh, I think that text should have been removed. I've reorganized the sections related to scoping rules and removed the rules that were incorrect. RPITIT should now be consistent with RPIT in inherent methods.

I also noticed there's an inconsistency in how the implementation handles this; see https://github.com/rust-lang/rust/issues/112194.

> Additionally, in the capture rules, the text talks only about generic parameters of the trait implementation - it says nothing about the trait definition.

We are talking about the hidden type supplied by each implementation. The type supplied by the implementation can't name the parameters of the trait itself because those aren't in scope – instead, usually you would have a generic parameter of the impl that _corresponds_ to each trait parameter. (I realize this is a bit subtle, but in the reference-level sections it helps to be pedantic.)

> ```rust
> trait Trait<'a> {
>     fn test() -> impl Sized;
> }
> ```

In this example the return type is not allowed to capture `'a` because it does not appear in `impl Sized`. On the other hand, if there were a type parameter in scope, the concrete type returned by our impl could capture that type parameter (and therefore would refer to any lifetimes the actual type for that parameter includes).

aliemjay commented:
> It is not clear to me, for example, whether impl Sized captures the lifetime 'a here:
> ```rust
> trait Trait<'a> {
>     fn test() -> impl Sized;
> }
>```

The decision here doesn't seem to be straightforward...

If it captures the lifetime `'a`, we would have an inconsistency with inherent impls. If it doesn't, we should treat the types of `<T as Trait<'a>>::test()` and `<T as Trait<'b>>::test()` to be equal for two different lifetimes. This might seem reasonable but it contradicts with the desugaring specified in this RFC because projection types are generic over all trait inputs, i.e. `<T as Trait<'a>>::Opaque#0` and `<T as Trait<'b>>::Opaque#0` are not equal types unless `'a == 'b`. For a concrete example, see how this example fails with the current implementation:
```rust
#![feature(return_position_impl_trait_in_trait)]

trait Trait<'a> {
    fn test() -> impl Sized;
}

fn test<'a, 'b, T: for<'x> Trait<'x>>() {
    let mut x;
    x = <T as Trait<'a>>::test();
    x = <T as Trait<'b>>::test();
    //^~ ERROR `'a` and `'b` must be the same: replace one with the other
}
```

Consequently we should change the desugaring to involve a supertrait that does not have the lifetime `'a` in its parameters:
```rust
trait SuperTrait {
    type AnonOpaque;
}

trait Trait<'a>: SuperTrait {
    fn test() -> Self::AnonOpaque;
}
```

tmandry commented:
> `impl Trait` in trait functions with default body is underspecified. Is it allowed? If yes, we should specify the desugaring. It is not trivial as it depends on the feature of associated type defaults. We should also mention the drawback of not being able to apply `#[refine]` to it.

Yes it is; I've updated the text to clarify that. I don't think there are any nonobvious interactions with associated type defaults, given that the types cannot be named in the first place.

Not being able to apply `#[refine]` is an existing drawback of that feature, not something specific to this RFC.

> The text specifies the legal positions in trait definitions but leaves trait implementations unspecified. Do we allow this?

Elaborated what's allowed in the RFC. The short answer is no, but it might be an interesting future possibility.

> Should we limit the legal positions for `impl Trait` to be maximally compatible with the naming schemes for return types?

My initial thought to this was no, but I think it's worth pondering. Added this as an unresolved question.

> The decision here doesn't seem to be straightforward...

I think the decision is captured as an unresolved question around https://github.com/rust-lang/rust/issues/112194. I updated the RFC text to point to your comment as well.

tmandry commented on text/0000-return-position-impl-trait-in-traits.md:
```diff
@@ -0,0 +1,679 @@
+
+[scoping]: https://rust-lang.github.io/rfcs/1951-expand-impl-trait.html#scoping-for-type-and-lifetime-parameters
+
+Lifetime parameters not in scope may still be indirectly named by one of the type parameters in scope.
```
The intent is to be consistent with inherent methods, but the implementation is not currently. I believe this is captured in the unresolved question around https://github.com/rust-lang/rust/issues/112194.

nikomatsakis commented:
**Huzzah!** The @rust-lang/lang team has decided **to accept** this RFC. 

Tracking issue: https://github.com/rust-lang/rust/issues/91611

If you'd like to keep following the development of this feature, please subscribe to that issue, thanks! :)

nikomatsakis commented:
(@aliemjay -- my sense was that the specific points you raised were (a) important details to be resolved but not reasons not to accept the RFC and (b) either captured in unresolved questions or clarified; if you feel otherwise, please follow-up either on tracking issue or zulip)
